### DISCUSSION TITLE
X509 getExtension is critical?

### DISCUSSION THREAD

**Discussion by ToastStudios** [phpseclib/phpseclib#2036] [Q&A] [answered]
My goal is to get all request extensions and whether they are are critical or not.
At the moment I iterate over `getExtensions` to get requested extensions with `getExtension`, but there is no info about the critical state. Am I missing something?

```
$x509 = new X509();
$x509->loadCSR($this->pkcs10()); // load the csr
foreach ($x509->getExtensions() as $extension) {
    echo($extension.':'.PHP_EOL);
    var_dump($x509->getExtension($extension));
}
```

**Comment by terrafrost** [comment, marked as answer]
For the time being I'd suggest doing something like this instead:

```php
$csr = $x509->loadCSR('...');

foreach ($csr['certificationRequestInfo']['attributes'] as $attribute) {
    if ($attribute['type'] != 'pkcs-9-at-extensionRequest') {
        continue;
   }
    print_r($attribute['value'][0]);
    break;
}
```
That said, I can add a `$x509->getExtensionCriticalStatus($id)` method if you'd like! LMK!

**Reply by ToastStudios** [reply]
This is exactly what I was looking for, thank you!
An extra method would still be great for readability, until then I use your workaround.

**Comment by terrafrost** [comment]
I added a new function - `getRequestedCertificateExtensions` - to X509.php:

https://github.com/phpseclib/phpseclib/commit/c228cc002b69644ed6871b4c3d8ecf50b19330c7#diff-ed108d6dcb12db1d73ce64c9c23d9bda034fd74be9b099617a8bf84fa72945b8R3512

I didn't add `$x509->getExtensionCriticalStatus($id)` because that doesn't mesh super well with the existent `$x509->getExtension()` method, which works differently. So with `getRequestedCertificateExtensions` you can just loop through all the requested extensions and see which of them are critical and which aren't.
